KOGITO-3654: Support FEEL code completion and code highlighting on all FEEL based cells

karreiro opened the pull request:
- [KOGITO-3653](https://issues.redhat.com/browse/KOGITO-3653): Support FEEL code highlight

- [KOGITO-3654](https://issues.redhat.com/browse/KOGITO-3654): Support FEEL code completion on all FEEL based cells

- Live demo: http://karreiro.com/kogito-editors-java

---

<img width="1054" alt="Screenshot 2021-07-29 at 14 01 51" src="https://user-images.githubusercontent.com/1079279/127488647-163d4f6f-708b-4ad1-8264-28a5c14375f8.png">

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

karreiro commented:
@vpellegrino @jomarko The live demo of this PR is updated and it's ready for review :-)

karreiro commented:
Thank for the review, @jomarko!

Yeah, actually, it's expected. This PR adopts the same approach that we adopted in the old boxed expression editor for instantiating Monaco. So the same behavior is already present in the editor:

<img width=500 src="https://user-images.githubusercontent.com/1079279/128207762-37eed6ac-996c-4071-9101-9e2f0d2bef15.gif">

The keyword is losing the bold style because the editor considers it a FEEL function (which is wrong). Thus, I've raised [KOGITO-5657](https://issues.redhat.com/browse/KOGITO-5657), so we can fix it in a future PR :-)

vpellegrino commented on kogito-editors-js/packages/boxed-expression-component/src/__tests__/components/Table/Table.test.tsx:
```diff
@@ -299,13 +296,14 @@ describe("Table tests", () => {
 
       expect(mockedOnRowsUpdate).toHaveBeenCalled();
-      expect(mockedOnRowsUpdate).toHaveBeenCalledWith([newRow]);
+      expect(mockedOnRowsUpdate).toHaveBeenCalledWith([row]);
```
A question about this test case: should the callback be called with update row value (i.e. `row[columnName] = "new value\t"`)?

karreiro replied:
Fixed! Thanks :-)

vpellegrino commented on kogito-editors-js/packages/feel-input-component/showcase/.env:
```diff
@@ -0,0 +1 @@
+SKIP_PREFLIGHT_CHECK=true
```
I would add a new line at the end of file

karreiro replied:
Fixed as well! Thanks 🚀

vpellegrino reviewed (commented):
@karreiro just a couple of very minor comments.
The work overall is brilliant. I really liked the introduction of the `feel-input-component` mainly from modularity and from re-usage perspectives.
Thanks for it!
**Side note**: I was just wondering, from a dev point of view, if steps specified at `kogito-editors-java` level are enough for having, for example, the `boxed-expression-component` up and running or we should specify better in its `README.md` that before running it, it is necessary to run: `yarn boostrap` and `yarn build:dev` at the `kogito-editors-js` level.

@jomarko looking at the online editor it seems that the defect https://issues.redhat.com/browse/KOGITO-4799 have been solved, can you confirm it? Thanks!

karreiro commented:
Thank you for the review, @vpellegrino!

Regarding the side note, I agree. I believe we may add an extra tutorial to our docs section, in the future, with an example of dependent components. I think it would clarify things a lot :-)

Also, I just updated the PR fixing all issues listed here and that one you mentioned in the chat (related to the _Relations_)!

Thanks again for the review :-)

vpellegrino reviewed (approved):
Thanks @karreiro

karreiro commented:
(the branch was rebased to solve conflicts) 👀

jomarko commented:
@vpellegrino hello, sorry, I can not confirm KOGITO-4799 is fixed.

I rechecked KOGITO-4799, during KOGITO-3654 re-review. The current state is:

Cell height is auto-computed for:
- Decision Table
- Relation

Other expression types loose the text completely for multiline inputs. - is this caused by this PR?
![Screenshot from 2021-08-09 07-04-35](https://user-images.githubusercontent.com/8044780/128662186-ac2e8662-a496-4311-8e06-972449f74646.png)

jomarko commented:
@karreiro https://github.com/kiegroup/kogito-editors-java/pull/83#issuecomment-894954342 is a blocker from my point of view.

karreiro commented:
Thank you for for the review, @jomarko 🚀 it's a blocker indeed, checking it right now :-)

karreiro commented:
@jomarko, thanks again for the review! I just rebased the branch, fixed the issue, and updated the live demo (in the description) with a fixed version of the component.

jomarko commented on kogito-editors-js/packages/feel-input-component/jest.config.js:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
2021

karreiro replied:
Fixed! Thanks! :)

jomarko commented on kogito-editors-js/packages/feel-input-component/showcase/src/index.tsx:
```diff
@@ -0,0 +1,96 @@
+
+    window.clearTimeout(window.__KIE__FEEL__THROTTLING___);
+    window.__KIE__FEEL__THROTTLING___ = window.setTimeout(() => {
+      const FEEL_SERVER = "https://piscine-madame-68280.herokuapp.com";
```
I am not able to access this URL, why we use it please?

generally I think it is not an recommended approach to access third party URLs in your application.

Could you please explain a little bit more this url?

karreiro replied:
Well spotted, @jomarko. I just removed this in favor of an optional property.

As I mentioned in the other comment, we will have the FEEL output component shortly. However, our FEEL client-side-parser is not as fully stable as the server-side one.

This URL points to a backend fallback (that will rely on DMN runner) to evaluate FEEL expressions that the FEEL client-side cannot process.

jomarko commented on kogito-editors-js/packages/feel-input-component/src/__tests__/Monaco/FeelConfigs.test.ts:
```diff
@@ -0,0 +1,103 @@
+    expect(config.value).toBe(value);
+  });
+
+  test("feelDefaultSuggestions", () => {
```
if I compare the number of entries in `FeelConfig.ts` and the number of entries we cover by test here, I need to say we do not cover them all. How we are sure we configured monaco for all supported keywords, functions etc?

do we have some service that returns list of reserved keywords? shouldn't we compare number of entries returned by such a serive with number of entries in `FeelConfig.ts`?

karreiro replied:
@jomarko In the test assertions, we're checking if we have 47 entries, and we're also checking 2 samples of each type of suggestion. In other words, 2 samples of keyword suggestions and 2 samples of function suggestions.

With these assertions, I believe we already have a high level of confidence without too much duplication. However, if you think it's a good idea to add assertions for all 47 entries, we can add them. I don't think it would be harmful, but I personally think it would be a bit of duplication.

jomarko reviewed (commented):
thank you @karreiro . some comments/questions inline. Manual check passed.

Found also some files, I am not sure they should be under SCM control.

![Screenshot from 2021-08-10 07-19-59](https://user-images.githubusercontent.com/8044780/128813431-fae56241-4b36-4950-aa48-00ebf85b3c55.png)

karreiro commented:
Thank you for the review, @jomarko!

I'm glad you noticed those files (I should have mentioned them in the PR description).

The `feel-input-component` soon will support behaviors like this  [http://karreiro.com/feel](http://karreiro.com/feel) (noticed the FEEL output part).

The showcase already includes part of the initial implementation to ensure that the `FeelInput` API is ready to receive such features.

Those files are actually the FEEL client-side-parser (from `drools/kie-dmn`). The expectation is to have something `feel-parser-component` wrapping this into a JS module soon.

The `feel-input-component/showcase` web app is a bit more focused on ensuring that `FeelInput` has a good API. So, it's important to keep those files there, for now. But, as soon as we have that JS module for the FEEL parser, we will remove them in favor of that dependency.

The PR is updated and ready for review :-)

jomarko reviewed (approved):
thank you @karreiro
